bench/macro/lsm-tree-bench-unions.hs

```diff
@@ -0,0 +1,629 @@
+        -- hence no action in required.
+        Nothing           -> pure ()
+        -- When the seed was explicitly specified by the user,
+        -- overide the system RNG with the explict seed to ensure determinism.
+        Just explicitSeed -> Random.setStdGen $ Random.mkStdGen explicitSeed
+
+deriveMultipleRNGs :: Int -> IO (NonEmpty Random.StdGen)
+deriveMultipleRNGs amount | amount <= 0 = fail "Must specify a POSITIVE number of RNGs to request"
+deriveMultipleRNGs amount =
+   let splitContinuation 0 gens done = (done, gens)
+       splitContinuation n gens prev =
+              let (curr, next) = Random.splitGen prev
+              in  splitContinuation (n - 1) (curr <| gens) next
+   in  do  rng0 <- Random.getStdGen
+           let (rng1, rng2) = Random.splitGen rng0
+               (rngN, splitRNGs) = splitContinuation amount (rng1 :| []) rng2
+           Random.setStdGen rngN
+           pure splitRNGs
+
+
+-------------------------------------------------------------------------------
+-- Sampling generation & measurements
+-------------------------------------------------------------------------------
+
+-- | The tuple of measured timings over a sample series @(mean, min, max, variance)@.
+type TimingResult = (Integer, Integer, Integer, Integer)
+
+-- | A single batch to be sampled as part of a sampling series for a measurement.
+{-# INLINE generateBatchSampling #-}
+generateBatchSampling  ::
+       Int       -- ^ initial size of the collection
+    -> Int       -- ^ batch size
+    -> IO (V.Vector K)
+generateBatchSampling initialSize batchSize = toOperations <$> V.replicateM batchSize randomKey
+  where
+    randomKey :: IO Word64
+    randomKey = Random.randomRIO (0, 2 * fromIntegral initialSize - 1)
+
+-- | Generate operation inputs
+{-# INLINE toOperations #-}
+toOperations :: V.Vector Word64 -> V.Vector K
+toOperations lookups = batch1
+  where
+    batch1 :: V.Vector K
+    batch1 = V.map makeKey lookups
+
+-- | Measure a series of of batched lookups.
+measureUsingSampleSeries ::
+       V.Vector (V.Vector K)
+    -> LSM.Table K V
+    -> IO TimingResult
+measureUsingSampleSeries sampleSeries table =
+    forM sampleSeries (measureBatchedLookup table) <&> \sampledTimings ->
+        let !popSize = toInteger $ V.length sampleSeries
+            !timingMean = round $ sum sampledTimings % popSize
+            !timingMax  = maximum sampledTimings
+            !timingMin  = minimum sampledTimings
+            !timingVariance =
+                let diffMeanSquared =
+                        (\x -> let diff = (x - timingMean) in diff * diff) <$> sampledTimings
+                in  round $ sum diffMeanSquared % popSize
+        in  (timingMean, timingMax, timingMin, timingVariance)
+
+-- | Measure the time to perform the batch of lookups on the supplied table.
+{-# INLINE measureBatchedLookup #-}
+measureBatchedLookup :: LSM.Table K V ->  V.Vector K -> IO Integer
+measureBatchedLookup table = timed_ . void . LSM.lookups table
```

Comment:
Would it make sense to time the running of *all* batches instead of one by one? That would "smooth" out any variance automatically

Reply:
Maybe. Probably even likely. That is what was occurring originally. However, if it's all the same to you, I'd rather not go through the effort to revert it back.